paparazzi/src/main/java/app/cash/paparazzi/internal/apng/ApngReader.kt

```diff
@@ -0,0 +1,264 @@
+  }
+
+  private fun BufferedSource.readFDAT() {
+    val nextSequence = readInt()
+    if (sequenceNumber + 1 != nextSequence) {
+      throw IOException("Out of order sequence, expected: ${sequenceNumber + 1} actual: $nextSequence")
+    }
+    sequenceNumber = nextSequence
+    decompress().decodeToScratchSpace()
+  }
+
+  private fun BufferedSource.decompress(): Buffer {
+    val inflater = Inflater().apply {
+      setInput(readByteArray())
+    }
+
+    val buffer = ByteArray(BUFFER_SIZE)
+    return Buffer().apply {
+      do {
+        val byteCount = inflater.inflate(buffer)
+        write(buffer, 0, byteCount)
+      } while (!inflater.finished() && byteCount != 0)
+    }
+  }
+
+  private fun BufferedSource.decodeToScratchSpace() {
+    val frameWidth = frameInfo?.width ?: width
+    val frameHeight = frameInfo?.height ?: height
+    val frameOffsetX = frameInfo?.offsetX ?: 0
+    val frameOffsetY = frameInfo?.offsetY ?: 0
+
+    for (y in 0 until frameHeight) {
+      val filterByte = readByte()
+      if (filterByte != 0.toByte()) {
+        throw IOException("Only filter None is supported, $filterByte")
+      }
+
+      for (x in 0 until frameWidth) {
+        val rgba = if (hasAlpha) {
+          readInt()
+        } else {
+          (readByte().toInt() shl 24) or
+            (readByte().toInt() shl 16) or
+            (readByte().toInt() shl 8)
+        }
+
+        val argb = rgba.rotateRight(8)
+        val position = ((frameOffsetY + y) * width) + frameOffsetX + x
+        scratchSpace[position] = argb
+      }
+    }
+
+    frameNumber++
+  }
+
+  private fun BufferedSource.readChunk(): Pair<Header, BufferedSource> {
+    val dataLength = readInt()
+    val chunkId = readByteArray(4L)
+    val data = readByteArray(dataLength.toLong())
+    val crc = readInt()
+
+    crcEngine.reset()
+    crcEngine.update(chunkId, 0, 4)
+    if (dataLength > 0) crcEngine.update(data, 0, dataLength)
+
+    if (crcEngine.value.toInt() != crc) {
+      throw IOException("CRC Mismatch decoding ${chunkId.decodeToString()}, invalid data")
+    }
+
+    return Header.valueOf(chunkId.decodeToString().uppercase()) to Buffer().apply { write(data) }
+  }
+
+  private fun BufferedSource.peekNextChunkId() =
+    peek().run {
+      skip(4L) // Ignored Data Length
+      readByteArray(4L)
+    }
+
+  private fun getFrame(): BufferedImage {
+    val buffer = DataBufferInt(scratchSpace, scratchSpace.size)
+    val bandMasks = intArrayOf(0xFF0000, 0xFF00, 0xFF, -0x1000000)
+    val raster = Raster.createPackedRaster(buffer, width, height, width, bandMasks, null)
+    val cm = ColorModel.getRGBdefault()
+
+    return BufferedImage(cm, raster, cm.isAlphaPremultiplied, null)
+  }
+
+  private data class FrameInfo(
+    val width: Int,
+    val height: Int,
+    val offsetX: Int,
+    val offsetY: Int,
+    val delayNumerator: Short,
+    val delayDenominator: Short
+  )
+
+  companion object {
+    private const val BUFFER_SIZE = 1_024_000
```

Comment:
Is this value dictated by the spec or something? If so, does it have a better name? Not sure it makes sense to be a constant otherwise–it's only used in one place and has no semantic meaning. I would probably put `1000 * 1024` directly at the call-site with a comment.

Reply:
Not dictated by the spec, just an arbitrary size for the Deflater. Will move to call-site